### DISCUSSION TITLE
Is there a feature to get intermediate certificates automatically?

### DISCUSSION THREAD

**Discussion by malsuke** [curl/curl#14009] [Q&A] [answered]
## Question

When studying certificate mechanisms, I discovered that curl requests do not encounter problems even without intermediate certificates.

While web browsers automatically fetch intermediate certificates through AIA fetching. 
I thought curl had a similar function about intermediate certificates.
I've looked around various issues and it didn't look like the AIA functionality had been developed yet!
https://github.com/curl/curl/issues/2793

Why then does curl on my Mac still succeed with requests? I would appreciate any insights you might have.

```bash
$ curl --version
curl 8.6.0 (x86_64-apple-darwin23.0) libcurl/8.6.0 (SecureTransport) LibreSSL/3.3.6 zlib/1.2.12 nghttp2/1.61.0
Release-Date: 2024-01-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtsp smb smbs smtp smtps telnet tftp
Features: alt-svc AsynchDNS GSS-API HSTS HTTP2 HTTPS-proxy IPv6 Kerberos Largefile libz MultiSSL NTLM NTLM_WB SPNEGO SSL threadsafe UnixSockets
```

## What I've Tried

In the following, I will use example.com as an example, but actually, it refers to a different site which an intermediate certificate is not configured

### Tested some Software.

```
Verification in Chrome: Accessed without any issues.
Verification in Firefox: Accessed without any issues.
Certificate check with OpenSSL: Encountered a certificate error.
curl on MacBook: Successfully retrieved HTML without any issues.
curl on Ubuntu 20.04: Encountered a certificate error.
curl using Docker: Encountered a certificate error.
Python's requests library: Encountered a certificate error.
```

### Tested using a curl container in Docker.

```bash
$ docker run --rm curlimages/curl:8.8.0 https://www.example.com/
curl: (60) SSL certificate problem: unable to get local issuer certificate
More details here: https://curl.se/docs/sslcerts.html

curl failed to verify the legitimacy of the server and therefore could not
establish a secure connection to it. To learn more about this situation and
how to fix it, please visit the web page mentioned above.
```

### Tested using a OpenSSL command.

```bash
$ openssl s_client -connect www.example.com:443                                            
Connecting to xxxxxxxx
CONNECTED(00000005)
depth=0 CN=www.example.com
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN=www.example.com
verify error:num=21:unable to verify the first certificate
verify return:1
depth=0 CN=www.example.com
verify return:1
```

### Tested using a curl built with LibreSSL.

Built and tested curl with LibreSSL on Ubuntu
Thinking LibreSSL might be involved, I built and tested curl on Ubuntu.
However, since the SSL error recurred, it seems LibreSSL is not the cause.

**Comment by bagder** [comment, marked as answer]
The simple answer is complicated.

The curl command line executable built by Apple and shipped on macOS is intentionally *broken*: they insert a sort of "backdoor" that changes how curl works and that is what tricked you. See my [blog post](https://daniel.haxx.se/blog/2024/03/08/the-apple-curl-security-incident-12604/) on the topic.

Even when you provide a specific separate CA cert bundle to test with, the curl version you try falls back and uses Apple's internal one if it fails. Which is exactly what happened for you here. I claim this is a security problem in Apple's curl version.

If you use curl from MacPorts or Homebrew (or build your own), you can see that those versions will fail on these intermediate certs just like curl does everywhere where the backdoor is not installed.

**Reply by malsuke** [reply]
I didn't know such a backdoor in MacOS's curl. Thank you for the valuable information.

However, I'm not sure on which internal certificates are being used in backdoor of MacOS's curl. 
I thought the certificates in the keychain were being used, but since there was no corresponding intermediate certificate there, I think MacOS's curl must be referencing a different file or something. If you have any information about the path of certificate file that backdoor uses or anything else, I would appreciate it if you could let me know.

Thank you.

**Reply by bagder** [reply]
Since the command line works for you when it should not, it seems like the obvious explanation given that we know Apple does this trickery with this exact outcome.

I cannot find any other explanation.

**Comment by duxsco** [comment]
The relevant RFCs basically state that every certificate except for the trust anchor must be sent by the server:
- RFC 5246 "The Transport Layer Security (TLS) Protocol Version 1.2" (Source: https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.2):
> certificate_list
> This is a sequence (chain) of certificates. The sender's certificate MUST come first in the list. Each following certificate MUST directly certify the one preceding it. Because certificate validation requires that root keys be distributed independently, the **self-signed certificate that specifies the root certificate authority MAY be omitted from the chain**, under the assumption that the remote end must already possess it in order to validate it in any case.
- RFC 8446 "The Transport Layer Security (TLS) Protocol Version 1.3" (Source: https://datatracker.ietf.org/doc/html/rfc8446#section-4.4.2):
> The sender's certificate MUST come in the first CertificateEntry in the list. Each following certificate SHOULD directly certify the one immediately preceding it. Because certificate validation requires that trust anchors be distributed independently, **a certificate that specifies a trust anchor MAY be omitted from the chain**, provided that supported peers are known to possess any omitted certificates.

**Apple's hack fosters RFC incompliant setups (IMHO)**. Lots of professionals I work with use MacBooks. Perhaps, they are content with Apple's curl not throwing any errors - even for imcompelete certificate chains - and let their service run in such a broken state. That might be one of the explanations for broken setups I see over and over.

Here is an example (https://service.santander.de/webapps/apps/filialen) on how not to do it:

https://web.archive.org/web/20240625090315/https://www.ssllabs.com/ssltest/analyze.html?d=service.santander.de&hideResults=on

and

```
❯ testssl.sh --server-defaults service.santander.de
...
 Chain of trust               NOT ok (chain incomplete)
...
```
